File: addTwoNumbersLinkedList.java
import java.util.*;
// Approach 1- Using stacks
class Solution { // TC O(m+n) SC O(m+n)
    public Node addTwoNumbersLinkedList(Node l1, Node l2) {
        Stack<Integer> s1 = new Stack<Integer>();
        Stack<Integer> s2 = new Stack<Integer>();
        
        while(l1 != null) {
            s1.push(l1.data);
            l1 = l1.next;
        };
        while(l2 != null) {
            s2.push(l2.data);
            l2 = l2.next;
        }
        
        int totalSum = 0, carry = 0;
        Node ans = new Node(-1);
        while (!s1.empty() || !s2.empty()) {
            if (!s1.empty()) totalSum += s1.pop();
            if (!s2.empty()) totalSum += s2.pop();
            
            ans.data = totalSum % 10;
            carry = totalSum / 10;
            Node head = new Node(carry);
            head.next = ans;
            ans = head;
            totalSum = carry;
        }

        return carry == 0 ? ans.next: ans;
    }
}

// Approach 2- Reverse the given linked list
class Solution1{ // TC O(m+n) SC O(m+n)
    public Node reverseList(Node head) {
        Node prev = null, temp;
        while (head != null) {
            // Keep the next node
            temp = head.next;
            // Reverse the link
            head.next = prev;
            // Update the previous node and the current node.
            prev = head;
            head = temp;    
        }    
        return prev;
    }

    public Node addTwoNumbers(Node l1, Node l2) {
        Node r1 = reverseList(l1);
        Node r2 = reverseList(l2);
        
        int totalSum = 0, carry = 0;
        Node ans = new Node(-1);
        while (r1 != null || r2 != null) {
            if (r1 != null) totalSum += r1.data;
            if (r2 != null) totalSum += r2.data;
            
            ans.data = totalSum % 10;
            carry = totalSum / 10;
            Node head = new Node(carry);
            head.next = ans;
            ans = head;
            totalSum = carry;

            r1 = r1 != null ? r1.next : null;
            r2 = r2 != null ? r2.next : null;
        }

        return carry == 0 ? ans.next: ans;
    }
}

class addTwoNumbersLinkedList{
    public static void main(String[] args) {
        LinkedList ll=new LinkedList();
        int[] keys1={7,2,4,3};
        int[] keys2={5,6,4};
        LinkedList list1=ll.constructLinkedList(keys1);
        LinkedList list2=ll.constructLinkedList(keys2);
        ll.printListHead(list1.head);
        ll.printListHead(list2.head);
        Solution s= new Solution();
        Node result=s.addTwoNumbersLinkedList(list1.head, list2.head);
        ll.printListHead(result);
        Solution1 s1= new Solution1();
        Node result1=s.addTwoNumbersLinkedList(list1.head, list2.head);
        ll.printListHead(result1);
    }
}

File: addTwoNumbersLinkedList.py
from implementLinkedList import *
class Solution:
# Approach 1- Using stacks
    def addTwoNumbersLinkedList(self, l1, l2): # TC O(m+n) SC O(m+n)
        s1, s2=[],[]

        while l1 or l2:
            if l1:
                s1.append(l1.val)
                l1=l1.next 
            if l2:
                s2.append(l2.val)
                l2=l2.next 
        
        totalSum=0
        ans=Node()
        while s1 or s2:
            if s1: totalSum+=s1.pop()
            if s2: totalSum+=s2.pop()
        
            ans.val=totalSum%10
            carry=totalSum//10
            head=Node(carry)
            head.next=ans 
            ans=head 
            totalSum=carry
        
        return ans.next if carry==0 else ans

# Approach 2- Reverse the linked list
    def reverseList(self, head): # TC O(m+n) SC O(m+n)
        prev = None
        temp = None
        while head:
            # Keep the next node
            temp = head.next
            # Reverse the link
            head.next = prev
            # Update the previous node and the current node.
            prev = head
            head = temp
        return prev

    def addTwoNumbers(self, l1, l2):
        r1 = self.reverseList(l1)
        r2 = self.reverseList(l2)

        total_sum = 0
        carry = 0
        ans = ListNode()
        while r1 or r2:
            if r1:
                total_sum += r1.val
                r1 = r1.next
            if r2:
                total_sum += r2.val
                r2 = r2.next

            ans.val = total_sum % 10
            carry = total_sum // 10
            head = ListNode(carry)
            head.next = ans
            ans = head
            total_sum = carry

        return ans.next if carry == 0 else ans

ll=LinkedList()
keys1=[7,2,4,3]
keys2=[5,6,4]
list1=ll.constructLinkedList(keys1)
list2=ll.constructLinkedList(keys2)
ll.printListHead(list1.head.next)
ll.printListHead(list2.head.next)
s=Solution()
result=s.addTwoNumbersLinkedList(list1.head, list2.head)
ll.printListHead(result)
        